eventsourcing/src/main/java/org/axonframework/eventsourcing/AnnotationBasedEventStateApplier.java

```diff
@@ -0,0 +1,136 @@
+     *                                 {@link org.axonframework.messaging.MessageType types} for
+     *                                 {@link org.axonframework.eventhandling.EventMessage EventMessages}
+     */
+    private AnnotationBasedEventStateApplier(
+            @Nonnull Class<M> modelType,
+            @Nonnull ParameterResolverFactory parameterResolverFactory,
+            @Nonnull MessageTypeResolver messageTypeResolver) {
+        this(parameterResolverFactory,
+             ClasspathHandlerDefinition.forClass(modelType),
+             messageTypeResolver);
+    }
+
+    /**
+     * Initialize a new {@link AnnotationBasedEventStateApplier}.
+     *
+     * @param parameterResolverFactory the strategy for resolving handler method parameter values
+     * @param handlerDefinition        the handler definition used to create concrete handlers
+     * @param messageTypeResolver      The {@link MessageTypeResolver} resolving the
+     *                                 {@link org.axonframework.messaging.MessageType types} for
+     *                                 {@link org.axonframework.eventhandling.EventMessage EventMessages}
+     */
+    public AnnotationBasedEventStateApplier(@Nonnull ParameterResolverFactory parameterResolverFactory,
+                                            @Nonnull HandlerDefinition handlerDefinition,
+                                            @Nonnull MessageTypeResolver messageTypeResolver) {
+        requireNonNull(parameterResolverFactory, "ParameterResolverFactory may not be null");
+        requireNonNull(handlerDefinition, "HandlerDefinition may not be null");
+        requireNonNull(messageTypeResolver, "MessageTypeResolver may not be null");
+        this.parameterResolverFactory = parameterResolverFactory;
+        this.handlerDefinition = handlerDefinition;
+        this.messageTypeResolver = messageTypeResolver;
+    }
+
+    @Override
+    public M apply(M model, @Nonnull EventMessage<?> event) {
+        requireNonNull(model, "Model may not be null");
+        requireNonNull(event, "Event Message may not be null");
+
+        // todo: should we create a new instance of AnnotationEventHandlerAdapter for each event?
+        // any idea how to cache that?
+        var handlerAdapter = new AnnotationEventHandlerAdapter(
+                model,
+                parameterResolverFactory,
+                handlerDefinition,
+                messageTypeResolver
+        );
+        try {
+            if (handlerAdapter.canHandle(event)) {
+                handlerAdapter.handleSync(event);
+            }
+        } catch (Exception e) {
+            // todo: I'm not sure about that, should we add Exception to the method signature and do not handle here?
+            throw new StateEvolvingException(
+                    "Failed to apply event [" + event.type() + "] in order to evolve [" + model.getClass() + "] state",
+                    e);
+        }
+        return model;
```

Comment:
Should we handle exception here? If not, should I add `throws Exception` to the method declaration in the EventStateApplier interface?